### PR TITLE
Add __iadd__ method to PauliSentence

### PR DESCRIPTION
**Pull request by AnuravModak**
So here is the first draft of the changes that i have done: towards #4658 

So for the task of defining def __ iadd__(self, other):
1) The iadd method takes two arguments, self and other.

2) It checks which of the two sentences (self or other) is smaller in terms of the number of terms.

3) It iterates over the smaller sentence, adding its terms and coefficients to the larger sentence in-place, without using the copy.

4) The method returns self to indicate in-place modification.

And for testing part, I have tried to optimize the code as far as possible, and from my understanding it should cover all the conventional test cases, please let me know if there is any loop hole or if something is not wrong.

### CODE REVIEW

**Comment by albi3ro** [comment]
Thanks for opening this PR @AnuravModak .  Just realized a potential clarification. If `__iadd__` is called, we should only add the terms to `self` and leave `other` unchanged. It will be up to the user to add terms to the bigger object.

For example if we have:
```
a = [1]
b = [2, 3]
a += b
```

`b` is left unchanged, while `a` is modified.

**Comment by AnuravModak** [comment]
so here i have made another PR. The idea was to only change the self and the other remains unchanged and the same has been reflected in `def test_iadd(self, string1, string2, result)`. Please review it and let me know if there is anything else to do.

**Comment by albi3ro** [comment]
Suppose we have two pauli sentences we want to sum together:
```
pw1 = PauliWord({0:"X", 1:"Y", 2:"Z"})
pw2 = PauliWord({1:"Y"})
pw3 = PauliWord({0:"Z"})

ps1 = PauliSentence({pw1: 1.0, pw2: 1.0})
ps2 = PauliSentence({pw1: 1.0, pw2: 1.0, pw3: 1.0})
```

We can use the normal addition:
```
>>> ps1 + ps2
2.0 * X(0) @ Y(1) @ Z(2)
+ 2.0 * Y(1)
+ 1.0 * Z(0)
```

Or the in-place addition on the larger sentence:
```
>>> c_ps2 = copy.copy(ps2)
>>> c_ps2 += ps1
>>> c_ps2
2.0 * X(0) @ Y(1) @ Z(2)
+ 2.0 * Y(1)
+ 1.0 * Z(0)
```

And get the same answer, as expected.

But what happens if we try and add the larger sentence onto the smaller one right now?

```
>>> c_ps1 = copy.copy(ps1)
>>> c_ps1 += ps2
>>> c_ps1
2.0 * X(0) @ Y(1) @ Z(2)
+ 2.0 * Y(1)
```

We end up getting `ps1 + ps1`, not `ps1 + ps2`.  

What does this behaviour tell us about what is going on? When we are adding a bigger pauli sentence to a smaller one, we are adding `self` to the object, instead of `other`. To get the correct answer, we always need to add `other`, regardless of size.

Does that make sense?

**Comment by AnuravModak** [comment]
Thanks for your clarification, the behavior you want to achieve, as explained in the provided text, is to always add the other (string2 in your test) to the self (string1 in your test) when performing in-place addition using +=. Currently, your __iadd__ method is designed to add the smaller of the two objects to the larger one, which can lead to the undesired behavior you mentioned.

So I am adding the draft code so please confirm if u are looking for something like this:

<pre>

def iadd(self, other):
    for key in other:
        if key in self:
            self[key] += other[key]
        else:
            self[key] = other[key]
    return self
</pre>

File: pennylane/pauli/pauli_arithmetic.py [hunk 1]
```diff
@@ -342,6 +342,15 @@ def __add__(self, other):
             larger_ps[key] += smaller_ps[key]
 
         return larger_ps
+
```
Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:345]:
```suggestion

```

File: tests/pauli/test_pauli_arithmetic.py [hunk 2]
```diff
@@ -409,6 +409,16 @@ def test_add(self, string1, string2, result):
         (ps5, "Can't get the matrix of an empty PauliSentence."),
     )
 
+    @pytest.mark.parametrize("string1, string2, result", tup_ps_add)
+    def test_iadd(self, string1, string2, result):
+        """Test that the correct result of inplace addition is produced and other object is not changed."""
+        original_string2=copy(string2)
```
[the remark this answers is not in the document]
Comment (AnuravModak) [reply tests/pauli/test_pauli_arithmetic.py:415]: Should I commit this suggestion within the current commit, or should I create a new commit separately to maintain a clear and organized commit history?

[the remark this answers is not in the document]
Comment (albi3ro) [reply tests/pauli/test_pauli_arithmetic.py:415]: Either works, so whatever works best for you. I often commit suggestions from the browser.

[the remark this answers is not in the document]
Comment (albi3ro) [reply tests/pauli/test_pauli_arithmetic.py:415]: Adding copies for both `string1` and `string2` made the tests pass locally for me.  Though I don't think the suggested change above is quite sufficient, as names would need ot be changed around a little.

**Review by albi3ro — COMMENTED** [review]
Once we get the test passing (pending copies in the right place so we don't mutate the test inputs), we will also need to add a changelog entry.

This is located at `doc/releases/changelog-dev.md`.  You should add an entry under `Improvements` and add your name to the list of contributors at the bottom.

Note that we will be creating our release candidate today, so the `changelog-dev.md` on master will be getting changed in a few hours.  So it might be best to wait a few hours to add a changelog entry to avoid merge conflict issues.

File: tests/pauli/test_pauli_arithmetic.py [hunk 3]
```diff
@@ -409,6 +409,16 @@ def test_add(self, string1, string2, result):
...
+    @pytest.mark.parametrize("string1, string2, result", tup_ps_add)
+    def test_iadd(self, string1, string2, result):
+        """Test that the correct result of inplace addition is produced and other object is not changed."""
+        original_string1=copy(string1)
+        string1 += string2
+        string1.simplify()
+
+        assert string1 == result # Check if the modified object matches the expected result
+        assert original_string2 == string2  # Ensure the original object is not modified
+
```
Comment (albi3ro) [inline tests/pauli/test_pauli_arithmetic.py:421]:
```suggestion
    @pytest.mark.parametrize("string1, string2, result", tup_ps_add)
    def test_iadd(self, string1, string2, result):
        """Test that the correct result of inplace addition is produced and other object is not changed."""
        copied_string1 = copy(string1)
        copied_string2 = copy(string2)
        copied_string1 += copied_string2
        copied_string1.simplify()

        assert copied_string1 == result # Check if the modified object matches the expected result
        assert copied_string2 == string2  # Ensure the original object is not modified
```

File: pennylane/pauli/pauli_arithmetic.py [hunk 4]
```diff
@@ -342,6 +342,14 @@ def __add__(self, other):
             larger_ps[key] += smaller_ps[key]
 
         return larger_ps
+    def __iadd__(self, other):
```
Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:345]:
```suggestion

    def __iadd__(self, other):
```

File: tests/pauli/test_pauli_arithmetic.py [hunk 5]
```diff
@@ -409,6 +409,16 @@ def test_add(self, string1, string2, result):
...
+    @pytest.mark.parametrize("string1, string2, result", tup_ps_add)
+    def test_iadd(self, string1, string2, result):
+        """Test that the correct result of inplace addition is produced and other object is not changed."""
+        copied_string1 = copy(string1)
+        copied_string2 = copy(string2)
+        copied_string1 += copied_string2
+        copied_string1.simplify()
+
+        assert copied_string1 == result # Check if the modified object matches the expected result
+        assert copied_string2 == string2  # Ensure the original object is not modified
```
Comment (albi3ro) [inline tests/pauli/test_pauli_arithmetic.py:421]:
```suggestion
        assert copied_string2 == string2  # Ensure the original object is not modified

```

**Comment by AnuravModak** [comment]
Hey @albi3ro, the Formatting check / black (pull_request) is still failing, is there anything to do from my side?

File: tests/pauli/test_pauli_arithmetic.py [hunk 6]
```diff
@@ -409,6 +409,17 @@ def test_add(self, string1, string2, result):
...
```
Comment (albi3ro) [inline tests/pauli/test_pauli_arithmetic.py:421]:
```suggestion
        assert copied_string2 == string2 # Ensure the original object is not modified
```

hopefully the last one 🤞 black can sometimes be really finicky

File: tests/pauli/test_pauli_arithmetic.py [hunk 7]
```diff
@@ -409,6 +409,17 @@ def test_add(self, string1, string2, result):
...
+    @pytest.mark.parametrize("string1, string2, result", tup_ps_add)
+    def test_iadd(self, string1, string2, result):
+        """Test that the correct result of inplace addition is produced and other object is not changed."""
+        copied_string1 = copy(string1)
+        copied_string2 = copy(string2)
+        copied_string1 += copied_string2
+        copied_string1.simplify()
+
+        assert copied_string1 == result # Check if the modified object matches the expected result
+        assert copied_string2 == string2 # Ensure the original object is not modified
```
Comment (albi3ro) [inline tests/pauli/test_pauli_arithmetic.py:421]: oops. it was wanting too spaces, not complaining about there being two spaces.
```suggestion
        assert copied_string1 == result  # Check if the modified object matches the expected result
        assert copied_string2 == string2  # Ensure the original object is not modified
```

downloading black: https://pypi.org/project/black/ and running
```
black -l 100 tests
```

makes the relevant changes. I was then comparing the git diff locally.

**Review by albi3ro — APPROVED** [review]
Thanks for your work and patience @AnuravModak 🎉 

I'm happy to approve, and am tagging the team for the second review.

**Review by timmysilv — APPROVED** [review]
looks good to me! 🎉

**Comment by trbromley** [comment]
Thanks @AnuravModak for your contribution! :trophy:

This should be released as part of v0.34 in early January. If you would like us to tag you as part of our Twitter/X marketing, feel free to optionally share your handle here.

**Comment by AnuravModak** [comment]
> Thanks @AnuravModak for your contribution! 🏆
> 
> This should be released as part of v0.34 in early January. If you would like us to tag you as part of our Twitter/X marketing, feel free to optionally share your handle here.

You can find me on twitter using this handle: @aakhri_modak.

**Pull request merged**